spec/abci++/abci++_methods_002_draft.md

```diff
@@ -0,0 +1,849 @@
+    | reject_senders | repeated string     | Reject the given P2P senders, regardless of `Result`. Any chunks already applied will not be refetched unless explicitly requested, but queued chunks from these senders will be discarded, and new chunks or other snapshots rejected. | 3            |
+
+```proto
+  enum Result {
+    UNKNOWN         = 0;  // Unknown result, abort all snapshot restoration
+    ACCEPT          = 1;  // The chunk was accepted.
+    ABORT           = 2;  // Abort snapshot restoration, and don't try any other snapshots.
+    RETRY           = 3;  // Reapply this chunk, combine with `RefetchChunks` and `RejectSenders` as appropriate.
+    RETRY_SNAPSHOT  = 4;  // Restart this snapshot from `OfferSnapshot`, reusing chunks unless instructed otherwise.
+    REJECT_SNAPSHOT = 5;  // Reject this snapshot, try a different one.
+  }
+```
+
+* **Usage**:
+    * The application can choose to refetch chunks and/or ban P2P peers as appropriate. Tendermint
+    will not do this unless instructed by the application.
+    * The application may want to verify each chunk, e.g. by attaching chunk hashes in
+    `Snapshot.Metadata` and/or incrementally verifying contents against `AppHash`.
+    * When all chunks have been accepted, Tendermint will make an ABCI `Info` call to verify that
+    `LastBlockAppHash` and `LastBlockHeight` matches the expected values, and record the
+    `AppVersion` in the node state. It then switches to fast sync or consensus and joins the
+    network.
+    * If Tendermint is unable to retrieve the next chunk after some time (e.g. because no suitable
+    peers are available), it will reject the snapshot and try a different one via `OfferSnapshot`.
+    The application should be prepared to reset and accept it or abort as appropriate.
+
+## New methods introduced in ABCI++
+
+### PrepareProposal
+
+#### Parameters and Types
+
+* **Request**:
+
+    | Name                    | Type                                        | Description                                                                                                      | Field Number |
+    |-------------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                    | bytes                                       | The block header's hash of the block to propose. Present for convenience (can be derived from the block header). | 1            |
+    | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                              | 2            |
```

Comment:
This is a bit difficult in `same-block` execution as we have fields that we currently don't know until we get a response from the proposer such as the `nextvalidatorhash` or the `apphash`. I'm not sure what the best way to go about it is.

Reply:
The reason for the header to be made available here is for the case where we have "next-block" execution, but we still want to fully execute the block here (e.g., to make 100% sure the block to propose is correct).
In this case, we need to provide the same data as `FinalizeBlock` does.
In contrast, in the same-block execution case, the App should be aware that this is the "header of the raw block", which the App has to turn into a "prepared block". Once the App does that and provides the response data (transactions, tx results, app_hash, etc.), Tendermint will re-calculate the block's header... which will be made available at `Finalizeblock` if this block is finally decided.
Does this clarify? Can you still see a problem here?

Reply:
I'm not too sure if we need to send the entire header to the application in the first place. I had a quick look through the SDK and the only things they use are Time and Height. Which makes sense - the header should be an almost purely Tendermint specific struct

Reply:
I believe the same reasoning goes for `FinalizaBlock` too. I propose to defer this to the moment we get to this part of the actual code, then we decide and propagate the decision back to the spec here if needed.

Reply:
@sergio-mena Do you mind opening an issue on this in the spec repo just so we don't forget?

Reply:
Sure! Please check #388